File: apps/agent/services/run_findings.py
"""What a finished agent run found, as text — kept in memory so later runs can recall it.

A plan's steps cannot use each other's results (FR-46), so a planned `memory.write` "summary" is
written before the research it summarises exists. Three runs in a row (2026-09-25/26) saved such a
placeholder; recall then ranked one first (similarity 0.93) and the next run built on nothing. The
owner's call: save the run's ACTUAL findings when it completes, and tell the planner not to fake
them. This module is the digest; `handle_agent_run_completed` saves it.

Mirrors `client/src/utils/runFindings.js` rule for rule (the Collaborator "Continue" digest);
`tests/unit/test_run_findings.py` holds both to the same fixture. Change one, change the other.
"""
from __future__ import annotations

from typing import Any, Iterable

FINDINGS_MARKER = "\n\n--- Findings from the previous run (attached by Collaborator) ---\n"
MAX_FINDINGS_CHARS = 6000
TRUNCATED = "\n[truncated]"


def _clean(value: Any) -> str:
    return str(value if value is not None else "").strip()


def _is_telemetry(node: Any) -> bool:
    return isinstance(node, dict) and str(node.get("source") or "").startswith("system_event:")


def step_findings(tool_name: str, result: Any) -> str | None:
    """One step's findings as text, or None when the step found nothing worth keeping."""
    if not isinstance(result, dict):
        return None
    if tool_name == "research.query":
        return _clean(result.get("raw_result")) or None
    if tool_name == "memory.recall":
        nodes = result.get("nodes") if isinstance(result.get("nodes"), list) else []
        notes = [n for n in nodes if isinstance(n, dict) and not _is_telemetry(n)]
        return "\n".join(f"- {_clean(n.get('content'))}" for n in notes) or None
    if tool_name == "search.query":
        hits = result.get("results") if isinstance(result.get("results"), list) else []
        hits = [h for h in hits if isinstance(h, dict) and (h.get("title") or h.get("snippet"))]
        return "\n".join(
            f"- {_clean(h.get('title'))}"
            + (f": {_clean(h.get('snippet'))}" if h.get("snippet") else "")
            + (f" ({h.get('url')})" if h.get("url") else "")
            for h in hits
        ) or None
    if tool_name == "arm.analyze":
        return _clean(result.get("summary")) or None
    if tool_name == "reasoning.evaluate":
        if not result.get("decision_type"):
            return None
        text = f"Recommendation: {result['decision_type']}"
        if result.get("reason"):
            text += f" (because {result['reason']})"
        if result.get("next_action_title"):
            text += f"\n{_clean(result['next_action_title'])}"
        return text
    if tool_name == "task.create":
        return f"Created task: {_clean(result['name'])}" if result.get("name") else None
    return None


def build_findings_digest(steps: Iterable[dict]) -> str:
    """Every successful step's findings, labelled by tool, capped at MAX_FINDINGS_CHARS.

    `steps` are dicts with `tool_name`, `status` and `result` — the shape the steps API returns.
    """
    sections = []
    for step in steps or []:
        if str(step.get("status") or "").lower() != "success":
            continue
        text = step_findings(str(step.get("tool_name") or ""), step.get("result"))
        if text:
            sections.append(f"[{step.get('tool_name')}]\n{text}")
    digest = "\n\n".join(sections)
    if len(digest) > MAX_FINDINGS_CHARS:
        return digest[:MAX_FINDINGS_CHARS] + TRUNCATED
    return digest


def goal_ask(goal: Any) -> str:
    """The owner's words from a goal, without any findings Collaborator attached."""
    text = str(goal or "")
    at = text.find(FINDINGS_MARKER)
    return text if at == -1 else text[:at]

File: client/src/utils/runFindings.js
import { safeMap } from "./safe";

// What a completed agent run found, as text the next run's planner can read.
//
// A plan's steps cannot use each other's results (FR-46): every argument is written before any
// step runs. On 2026-09-25/26 that meant "research X, then use it" planned its second half blind,
// and the owner carried the findings into the next goal by hand, copied out of the database.
// Collaborator does the carrying instead: it digests the finished run's step results and attaches
// them to the follow-up goal, visibly, so the owner approves exactly what the planner will see.

// Separates the owner's words from the attached findings inside a goal. Kept distinctive so a
// goal can be split back apart for display; the planner reads it as plain text.
export const FINDINGS_MARKER = "\n\n--- Findings from the previous run (attached by Collaborator) ---\n";

// A research result alone is ~2 KB. The cap keeps a follow-up goal a size the planner prompt can
// carry without crowding out the plan, KPI and tool context around it.
export const MAX_FINDINGS_CHARS = 6000;

// Recall ranks the user's own nodes, most of which are system telemetry ("execution.started from
// masterplan.pace.propose"); they are never a finding.
const isTelemetry = (node) => String(node?.source || "").startsWith("system_event:");

const clean = (s) => String(s ?? "").trim();

// One step's findings as text, or null when the step found nothing worth carrying.
export function stepFindings(step) {
  const r = step?.result;
  if (!r || typeof r !== "object") return null;
  switch (step.tool_name) {
    case "research.query":
      return clean(r.raw_result) || null;
    case "memory.recall": {
      const notes = (Array.isArray(r.nodes) ? r.nodes : []).filter((n) => !isTelemetry(n));
      return notes.length ? safeMap(notes, (n) => `- ${clean(n.content)}`).join("\n") : null;
    }
    case "search.query": {
      const hits = (Array.isArray(r.results) ? r.results : []).filter((h) => h?.title || h?.snippet);
      return hits.length
        ? safeMap(hits, (h) => `- ${clean(h.title)}${h.snippet ? `: ${clean(h.snippet)}` : ""}${h.url ? ` (${h.url})` : ""}`).join("\n")
        : null;
    }
    case "arm.analyze":
      return clean(r.summary) || null;
    case "reasoning.evaluate":
      return r.decision_type
        ? `Recommendation: ${r.decision_type}${r.reason ? ` (because ${r.reason})` : ""}` +
            (r.next_action_title ? `\n${clean(r.next_action_title)}` : "")
        : null;
    case "task.create":
      return r.name ? `Created task: ${clean(r.name)}` : null;
    default:
      return null;
  }
}

// Every successful step's findings, labelled by tool, capped at MAX_FINDINGS_CHARS.
export function buildFindingsDigest(steps) {
  const succeeded = (Array.isArray(steps) ? steps : []).filter(
    (s) => String(s?.status || "").toLowerCase() === "success"
  );
  const sections = safeMap(succeeded, (s) => {
    const text = stepFindings(s);
    return text ? `[${s.tool_name}]\n${text}` : null;
  }).filter(Boolean);
  const digest = sections.join("\n\n");
  return digest.length > MAX_FINDINGS_CHARS
    ? `${digest.slice(0, MAX_FINDINGS_CHARS)}\n[truncated]`
    : digest;
}

export function composeFollowUpGoal(nextGoal, digest) {
  const ask = clean(nextGoal);
  return digest ? `${ask}${FINDINGS_MARKER}${digest}` : ask;
}

// The owner's words, and whether findings were attached — for showing a goal without the digest.
export function splitGoal(goal) {
  const text = String(goal ?? "");
  const at = text.indexOf(FINDINGS_MARKER);
  return at === -1
    ? { ask: text, hasFindings: false }
    : { ask: text.slice(0, at), hasFindings: true };
}
